feat: run full nodes via docker (#38)
availproject/availproject.github.io: docs/node/full-node/full-node-docker.md

---
id: full-node-docker
title: How to Run a Full Node using Docker
sidebar_label: Run a Full Node Using Docker
sidebar_position: 2
description: "Learn how to run an Avail full node using Docker."
keywords:
  - docs
  - avail
  - node
  - full node
  - data availability
  - da
  - docker
image: https://availproject.github.io/img/avail/AvailDocs.png
---
import useBaseUrl from '@docusaurus/useBaseUrl';

The Avail team distributes official Docker images which can be used to run nodes on the **Kate testnet** network . These instructions are for running a Full Node, but they can be adapted for running validators as well.

> ## Before you start

Reviewer: ```suggestion
## Before You Start
```

The instructions below assume a Linux distribution with `apt` (Debian,
for example), but you may be able to adapt them to a different
distribution. Also the Docker images Currently Available are only supported for **Linux/amd64 or x86_64 based** cpus .It's also common for users to run nodes on a cloud
server. The following list of standard hardware is a recommendation of
hardware specs that your environment should have:

| Minimum                            | Recommended                         |
|------------------------------------|-------------------------------------|
| 4GB RAM                            | 8GB RAM                             |
| 2 core CPU amd64/x86 architecture  | 4 core CPU  amd64/x86 architecture  |
| 20-40 GB SSD                       | 200-300 GB SSD                      |

The peer exchange for a Avail full node generally depends on port 30333 being open. When you configure your firewall or security groups for Cloud Providers, make sure these ports are open along with whatever ports you need to access the machine.

#### Info Alternate networks & releases

> All instructions below are for the **Kate testnet** network. To join a

Reviewer: ```suggestion
The instructions provided are specifically for the **Kate testnet**.
```

> different network, you may need to download a different node version

Reviewer: ```suggestion
To connect to a different network, you may need to download an alternate node version
```

from the [node releases

> page](https://github.com/availproject/avail/releases) and a chain

Reviewer: ```suggestion
page](https://github.com/availproject/avail/releases) and acquire the corresponding
```

> specification file for the specific network. Contact the Avail team if

Reviewer: ```suggestion
chain specification file. For any queries, feel free to reach out to the Avail team.
```

> you have questions.

Reviewer: ```suggestion
```

> ## Useful Links

Author: i would keep this as separate section as there will be more additional links in future which will not be part of initial process

[Dockerhub Repository](https://hub.docker.com/r/availj/avail/tags)

[Github Releases](https://github.com/availproject/avail/releases)

## Initial Setup

At this point, you should have shell access with root privileges to a linux machine.

> ### Install Docker 

Reviewer: ```suggestion
### Install Docker

```

Most likely your operating system won’t have Docker installed by default. Please follow the instructions for your particular distribution found here: https://docs.docker.com/engine/install/

> We’re following the instructions for Ubuntu. The steps are included below, but please see the official instructions in case they’ve been updated.

Reviewer: ```suggestion
In this guide, we will use the Ubuntu-specific installation instructions. It's advisable to consult the official guidelines for the most up-to-date information.
```

```bash
sudo apt-get update
sudo apt-get install ca-certificates curl gnupg lsb-release
sudo mkdir -p /etc/apt/keyrings
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt-get update
sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose-plugin
```

> At this point you should have Docker installed.

Reviewer: ```suggestion
At this point, you should have Docker installed.
```

> In many cases, it’s inconvenient to run docker as **`root`** user so we’ll follow the post install steps [here](https://docs.docker.com/engine/install/linux-postinstall/) in order to interact with docker without needing to be **`root:`**

Reviewer: ```suggestion
To avoid running Docker as the **root** user, which is often inconvenient, we'll adhere to the post-installation steps outlined [here](https://docs.docker.com/engine/install/linux-postinstall/). This allows us to interact with Docker without requiring **root** privileges.
```

```bash
sudo groupadd docker
sudo usermod -aG docker $USER

```

> Now you should be able to logout and log back in and run docker commands without **`sudo.`**

Reviewer: ```suggestion
You should now be able to log out and log back in, and execute Docker commands without the need for **sudo**.
```

> ## Disk Setup

Reviewer: ```suggestion

## Disk Setup
```

> The exact steps required here are going to vary a lot based on your needs. Most likely you’ll have a root partition running your operating system on one device. You’ll probably want one or more devices for actually holding the blockchain data. For the rest of the walkthrough, we’re going to have that additional device mounted at **`/mnt/avail`**.

Reviewer: ```suggestion
The specific steps will vary significantly based on your requirements. Typically, you'll have a root partition for the operating system on one device, and one or more separate devices for storing blockchain data. For the remainder of this guide, we'll assume that the additional storage device is mounted at `/mnt/avail`.
```

> Before Mounting the Additional Disk it is recommended to format that disk for Use and Create a Filesystem You can [Follow Here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-using-volumes.html) for that.

Reviewer: ```suggestion
Before mounting the additional disk, it's advisable to format it and create a filesystem. For guidance on this process, you can [follow these instructions](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-using-volumes.html).
```

> In this example, we have a device with 300 GB of available space located at **`/dev/nvme1n1`**. We are going to mount that using the steps below:

Reviewer: ```suggestion
In this example, we're working with a 300 GB device located at `/dev/nvme1n1`. The steps to mount this device are outlined below:
```

```bash
sudo mkdir /mnt/avail
sudo mount -a /dev/nvme1n1 /mnt/avail

> ```

Reviewer: ```suggestion
```

```

We use **`df -h`** to make sure the mount looks good.

If that all looks good, we might as well create Additional Sub-Directories for our DATA and config store

```bash
sudo mkdir /mnt/avail/config
sudo mkdir /mnt/avail/state
sudo mkdir /mnt/avail/keystore
```

Depending on your use case and operating system, you’ll likely want to create an entry in **`/etc/fstab`** in order to make sure your device is mounted when the system reboots.

In our case we're following some steps like this:

```bash
# Use blkid to get the UUID for the device that we're mounting

blkid

# Edit the fstab file  and add a line to mount your device
# UUID={your uuid}      /mnt/data   {your filesystem}   defaults    0   1
sudo emacs /etc/fstab

#you can use any test editor based on your prefereance to edit the file here we have used emacs .

# use this to verify the fstab actually works
sudo findmnt --verify --verbose
```

> At this point you should be able to reboot and confirm that the system loads your mount properly.

Reviewer: ```suggestion

At this stage, you should be able to reboot and verify that the system correctly mounts your device upon startup.
```

## Avail setup

Download the Correct Chaispec file for the network in our case we are using the **`Kate-Testnet`** Chainspec.

   
   | Network      |Docker Hub |  Chain Specification File | Chain Info Reference|
   |--------------|-----------|--------------------------|----------------------|
   | Kate Testnet |[Releases](https://hub.docker.com/r/availj/avail/tags)|[chainspec.raw.json](/configs/kate/chainspec.raw.json) | [Chain Info](/configs/kate/chaininfo.txt) |

In our case we will download the chainspec into our config folder as `kate-chainspec.raw.json`

```bash
curl -L -o /mnt/avail/config/kate-chainspec.raw.json https://raw.githubusercontent.com/availproject/availproject.github.io/main/static/configs/kate/chainspec.raw.json
```

> Now that we have Our Chainspec config Downlaoded Let's start our Avail Node.

Reviewer: ```suggestion

Now that we've downloaded our Chainspec configuration, let's proceed to launch our Avail Node.
```

```bash
cd /mnt/avail

sudo docker run -v $(pwd)/config/kate-chainspec.raw.json:/da/genesis/chainspec.raw.json -v $(pwd)/state:/da/state:rw -v $(pwd)/keystore:/da/keystore:rw -e DA_CHAIN=/da/genesis/chainspec.raw.json -e DA_NAME=kate-docker-avail-Node -p 0.0.0.0:30333:30333 -p 9615:9615 -p 9933:9933 -d --restart unless-stopped availj/avail:v1.6.2-rc1
```

> in the above command steps include:

Reviewer: ```suggestion
Now that we've downloaded our Chainspec configuration, let's proceed to launch our Avail Node. The steps in the command include:
```

- changing directory to `/mnt/avail` so we are in correct directory 

> - we have mounted the `/mnt/avail/config/kate-chainspec.raw.json` to `/da/genesis/chainspec.raw.json` of the docker container so that the chainspec is correct

Reviewer: ```suggestion
- Mounting `/mnt/avail/config/kate-chainspec.raw.json` to `/da/genesis/chainspec.raw.json` in the Docker container for accurate chainspec.
```

- `/mnt/avail/state` to `/da/state` and gave it permission to read and write both so that our da data is stored on the separate disk and is presistent in case of even the container crashes.
- `DA_CHAIN` suggest which chainspec file to use inside the container
- `DA_NAME` is Name of your node you can give if any name in our example we have given kate-docker-avail-Node

> - ports convention **`30333`** is used for public `p2p connection` **`9615`** is the `prometheus metrics endpoint port` and **`9933`** is the `http rpc port` if you also want to add `websocket` you add port **`9944`** in the docker command.

Reviewer: ```suggestion
- Utilizing port **`30333`** for public P2P connections, **`9615`** for the Prometheus metrics endpoint, and **`9933`** for the HTTP RPC port. For WebSocket, add port **`9944`**.
```

> - image we have used if from the availj docker hub repository

Reviewer: ```suggestion
- Using an image from the Avail Docker Hub repository.
```

2023-08-21 08:30:04 Running JSON-RPC HTTP server: addr=127.0.0.1:9933, allowed origins=["http://localhost:*", "http://127.0.0.1:*", "https://localhost:*", "https://127.0.0.1:*", "https://polkadot.js.org"]
2023-08-21 08:30:04 Running JSON-RPC WS server: addr=127.0.0.1:9944, allowed origins=["http://localhost:*", "http://127.0.0.1:*", "https://localhost:*", "https://127.0.0.1:*", "https://polkadot.js.org"]
2023-08-21 08:30:04 🏁 CPU score: 671.55 MiBs
2023-08-21 08:30:04 🏁 Memory score: 4.47 GiBs
2023-08-21 08:30:04 🏁 Disk score (seq. writes): 339.36 MiBs
2023-08-21 08:30:04 🏁 Disk score (rand. writes): 62.48 MiBs
2023-08-21 08:30:05 🔍 Discovered new external address for our node: /ip4/13.53.42.153/tcp/30333/ws/p2p/12D3KooWEdgyAtH8ZCU8ScTx1hx5NWh4gmDGNcedtLxrJ1htSeBe2023-08-21 08:30:09 ⚙️  Syncing, target=#326624 (15 peers), best: #9406 (0x875e…c887), finalized #9317 (0x37b6…28ff), ⬇ 321.9kiB/s ⬆ 30.1kiB/s
2023-08-21 08:30:14 ⚙️  Syncing 64.4 bps, target=#326624 (15 peers), best: #9728 (0xb4fe…e318), finalized #9317 (0x37b6…28ff), ⬇ 40.2kiB/s ⬆ 1.8kiB/s
```

It will also be listed on the [Avail
   Telemetry](http://telemetry.avail.tools/) site under the "Node

>    name" that appears in the node command output. Note that there are

Reviewer: ```suggestion
    Be sure to select the appropriate network tab at the top corresponding to the network you've joined.
```

>    network tabs at the top, select the one for the network you joined.

Reviewer: ```suggestion
```

> That's all! You are now running an Avail full node 🎉

Reviewer: ```suggestion
That's it! You're now successfully running an Avail full node 🎉
```
